### PR TITLE
[e2e] Update create simple product test

### PR DESCRIPTION
**Pull request by alvarothomas**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Add the following steps to the test that checks the creation of a simple product:
- add a description
- add a sale price
- add SKU
- include a purchase
- add an attribute (Color)

Closes [woocommerce/woocommerce-quality/issues/607](https://github.com/woocommerce/woocommerce-quality/issues/607)

### How to test the changes in this Pull Request:

Please run the _create simple product test_ as follows:

`pnpm test:e2e-pw ./tests/e2e-pw/tests/merchant/create-simple-product.spec.js`

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Updates the test of a virtual product creation functionality with further characteristics

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js [hunk 1]
```diff
@@ -4,6 +4,12 @@ const wcApi = require( '@woocommerce/woocommerce-rest-api' ).default;
 const virtualProductName = 'Virtual Product Name';
 const nonVirtualProductName = 'Non Virtual Product Name';
 const productPrice = '9.99';
+const salePrice = '6.99';
+const productDescription = 'Virtual product description.';
+const productSKU = '1234567890'
+const productPurchaseNote = 'Virtual product purchase note'
+const productAttribute = 'color'
+const productAttributeColor = 'red | white'
```
Comment (veljkho) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: There are no ending `;` for these lines.
I have prettier installed in my IDE which automatically solves code formatting on save. I also see other formatting issues down below but I will comment it separately.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js [hunk 2]
```diff
@@ -57,7 +63,33 @@ test.describe.serial( 'Add New Simple Product Page', () => {
...
+		await page.getByText( 'Advanced' ).click();
+		await page.locator( '#_purchase_note' ).fill( productPurchaseNote );
+
+		// Fill in a color as attribute
+		await page
+			.locator( '.attribute_tab' )
+			.getByRole( 'link', { name: 'Attributes' } ).click();
+		await page.getByPlaceholder ( 'f.e. size or color' ).fill( productAttribute );
+		await page.getByPlaceholder ( 'Enter some descriptive text. Use “|” to separate different values.' ).fill( productAttributeColor );
+		await page.keyboard.press('Enter');
```
Comment (veljkho) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: 87 and 88 are very long. 89 needs spacing around `Enter`. These are solvable by a prettier formatted.
Please let me know if you need any help around it.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js [hunk 3]
```diff
@@ -57,7 +63,33 @@ test.describe.serial( 'Add New Simple Product Page', () => {
...
+
+		// Fill in a product description
+		await page
+			.frameLocator( '#content_ifr' )
+			.locator( '.wp-editor' )
+			.fill( productDescription );
+
+		// Fill in SKU
+		await page.getByText( 'Inventory' ).click();
+		await page.locator( '#_sku' ).fill( productSKU );
```
Comment (Stojdza) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: I've followed @lanej0's [suggestion](https://github.com/woocommerce/woocommerce/pull/43710#discussion_r1454084102) and used `getBy*` instead of `locator` wherever possible. For the SKU form, you can try the same I did, per @veljkho's advice, [here](https://github.com/woocommerce/woocommerce/pull/43710/files#diff-4d8b4fa2d3fc09e284b17868562b01cf7bc2d5294477590b31e45f0f01277ebbR156), it should work 🤞

**Review by Stojdza — COMMENTED** [review]
Left one comment for the SKU field

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js [hunk 4]
```diff
@@ -57,7 +63,33 @@ test.describe.serial( 'Add New Simple Product Page', () => {
...
 		await page.locator( '#title' ).fill( virtualProductName );
 		await page.locator( '#_regular_price' ).fill( productPrice );
+		await page.locator( '#_sale_price' ).fill( salePrice );
 		await page.locator( '#_virtual' ).click();
+
+		// Fill in a product description
+		await page
+			.frameLocator( '#content_ifr' )
+			.locator( '.wp-editor' )
+			.fill( productDescription );
```
Comment (veljkho) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:111]: It didn't work for me locally and I realised why.. I had Text view open by default and I don't know why. I'm afraid someone when testing the suite locally could have the same issue or this may produce false-positive.

Wouldn't you mind adding clicking Visual first before filling product description?

```
await page.getByRole( 'button', { name: 'Visual', exact: true } ).first().click();
```

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js [hunk 5]
```diff
@@ -57,7 +63,33 @@ test.describe.serial( 'Add New Simple Product Page', () => {
...
+
+		// Fill in SKU
+		await page.getByText( 'Inventory' ).click();
+		await page.locator( '#_sku' ).fill( productSKU );
+
+		// Fill in purchase note
+		await page.getByText( 'Advanced' ).click();
+		await page.locator( '#_purchase_note' ).fill( productPurchaseNote );
```
Comment (veljkho) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: I think you can use here `getByLabel( 'Purchase note' ).fill( productPurchaseNote );` and that would work nicely.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js (hunk 3 above)
Comment (lanej0) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: This looks good @Stojdza.

The failing CI check you have is just because you have to add a changelog entry.  If you go to the `plugins/woocommerce` folder and run `pnpm --filter=plugin-woocommerce run changelog add` you can do that, commit it and it should resolved the failing check.

**Review by lanej0 — APPROVED** [review]

**Review by veljkho — COMMENTED** [review]
I forgot to submit my review @alvarothomas, here are my comments down below.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js (hunk 4 above)
Comment (rodelgc) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:111]: The locator also didn't work for me when the "Product description" field was set to "Text". Would be great to add a step to click on the "Visual" tab first to ensure robustness of the product description locator, as Veljko mentioned above. 

Alternatively, you may also use Playwright Locator's [or](https://playwright.dev/docs/api/class-locator#locator-or) command so that you will end up `or`ing two locators for the Product description textbox, and these `or`ed locators will work on both "Visual" and "Text" mode.

I think either of those 2 approaches would work.

Comment (rodelgc) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: It will also be good to use user-facing locators such as any of the `page.getBy*` ones, since this is one of the best practices that Playwright suggests [here](https://playwright.dev/docs/best-practices#prefer-user-facing-attributes-to-xpath-or-css-selectors).

As a tip, you can let Playwright pick the locator for you using [UI mode](https://playwright.dev/docs/test-ui-mode#opening-ui-mode), specifically its [locator picker](https://playwright.dev/docs/test-ui-mode#pick-locator). Although not entirely perfect, the locator picker defaults to giving you a user-facing locator you can use.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js (hunk 1 above)
Comment (rodelgc) [inline plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: I would suggest adding assertions for these values later on in the test.  So far, the only assertion we have on this test is that the "Product published" notification was displayed. Adding more assertions around these newly added details will make the test more, well... assertive 😅

**Review by rodelgc — DISMISSED** [review]
Hi @alvarothomas I'm done reviewing the PR and requested some changes. 

This PR has a merge conflict, so kindly resolve that first before we merge it.

Thanks for working on this issue BTW! This gives the `can create simple virtual product` test more coverage. And great job learning our testing framework quickly!

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js (hunk 1 above)
Comment (rodelgc) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: As a bit of a context, we use the [@wordpress/prettier-config](https://www.npmjs.com/package/@wordpress/prettier-config) package for formatting. When setting up the local testing environment, this package will be installed as part of the `pnpm install` command in the set up steps. So you'll just have to press `Shift + Option + F ( ⇧⌥F )` and WP's own prettier config will fix most of the formatting issues Veljko raised.

Comment (alvarothomas) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: Thank you for pointing this out @veljkho , after +2 years using Python only I need to get used to JS formatting.

Comment (alvarothomas) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: Also, I'll try with the Prettier extension too. Thanks!

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js (hunk 2 above)
Comment (alvarothomas) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: Tried Prettier to format those. Nice one!

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js (hunk 3 above)
Comment (alvarothomas) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: Thanks @lanej0, I've just added this via CLI.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js (hunk 5 above)
Comment (alvarothomas) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: Good idea, thank you for suggesting that.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js (hunk 4 above)
Comment (alvarothomas) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:111]: Nice one, thanks for suggesting to click Visual beforehand.

Comment (alvarothomas) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: Thanks for this, @rodelgc !! It took me a while to understand how PW UI mode could help but, once I've got it, I've been able to use the suggestions to replace a bunch od locators with `getBy` instead. Please let me know what do you think about these changes when you can.

File: plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js (hunk 1 above)
Comment (alvarothomas) [reply plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js]: Good idea @rodelgc , if I'm not mistaken, this aligns with my [comment - second point -](https://github.com/woocommerce/woocommerce-quality/issues/607#issuecomment-1898576008) in the related issue.

tl;dr: My impression is that this script automates the creation of a product, but there's no test afterwards to verify that the product was created with the expected values. I hope I'm explaining myself right, not 100% if this is what you meant here too.

Would you please mind sharing with me one example of assertion that can be used to verify any of those values?

**Comment by alvarothomas** [comment]
@rodelgc @veljkho I've updated the `can create simple virtual product` test with some asserts to verify the values in the product editor. If I'm not mistaken, the rest of the values are verified in the `can have a shopper add the simple non-virtual product to the cart` test too.

I'd need some assistance to resolve the conflict on the create-simple-product file. Would it be enough to just bring these in?

```
const productTag = 'nonVirtualTag';
const productCategory = 'nonVirtualCategory';
const productDescription = 'Description of a non-virtual product.';
const productDescriptionShort = 'Short description';
```

**Review by veljkho — APPROVED** [review]
I rebased trunk and fixed conflict. @alvarothomas 
Worked greatly for me when I executed test. 👍
